sparse/_coo/core.py

```diff
@@ -2258,6 +2258,112 @@ def asformat(self, format, compressed_axes=None):
 
         raise NotImplementedError("The given format is not supported.")
 
+    def diagonal(self, offset=0, axis1=0, axis2=1):
+        """
+        Extract diagonal from a COO array. The equivalent of :obj:`numpy.diagonal`.
+
+        Parameters
+        ----------
+        offset: int, optional
+            Offset of the diagonal from the main diagonal. Defaults to main diagonal (0).
+        axis1: int, optional
+            First axis from which the diagonals should be taken.  
+            Defaults to first axis (0).
+        axis2 : int, optional
+            Second axis from which the diagonals should be taken.  
+            Defaults to second axis (1).
+
+        Examples
+        --------
+        >>> import sparse
+        >>> x = sparse.as_coo(np.arange(9).reshape(3,3))
+        >>> x.diagonal().todense()
+        array([0,4,8])
+        >>> x.diagonal(offset=1).todense()
+        array([1,5])
+
+        >>> x = sparse.as_coo(np.arange(12).reshape((2,3,2)))
+        >>> x_diag = x.diagonal(axis1=0, axis2=2)
+        >>> x_diag.shape
+        (3, 2)
+        >>> x_diag.todense()
+        array([[ 0,  7],
+               [ 2,  9],
+               [ 4, 11]])
+
+        Returns
+        -------
+        out: COO
+            The result of the operation.
+
+        Raises
+        ------
+        ValueError
+            If a.shape[axis1] != a.shape[axis2]
+
+        See Also
+        --------
+        :obj:`numpy.diagonal`: NumPy equivalent function
+        """
+        if self.shape[axis1] != self.shape[axis2]:
+            raise ValueError('shape[axis1] != shape[axis2]')
+
+        diag_axes = [axis for axis in range(len(self.shape)) if axis != axis1 and axis != axis2] + [axis1]
+        diag_shape = [self.shape[axis] for axis in diag_axes]
+        diag_shape[-1] -= abs(offset)
+
+        diag_idx = [i for i in range(len(self.data)) if self.coords[axis1][i]+offset == self.coords[axis2][i]]
+
+        diag_coords = [self.coords[axis][diag_idx] for axis in diag_axes]
+        diag_data  = self.data[diag_idx]
+
+        return COO(diag_coords, diag_data, diag_shape)
+
+    def diagonalize(self, axis=0):
```

Comment:
Method -> Function here as well. Also need to mirror the NumPy API...

Reply:
As discussed in #288, there is no numpy equivalent of ```diagonalize```. Thereis ```np.diag``` but that only does the same thing for 1D and actually does the opposite for 2D. It does not accept 3D or higher. I thought calling it ```diagonalize``` was the best solution to not confuse

Reply:
That seems reasonable... But future code might not work if we have to change the signature to be in line with NumPy, so I think a warning is in order in the docs.